### DISCUSSION TITLE
Proposal: Get runtime logs from control API

### DISCUSSION THREAD

**Discussion by bdbai** [eycorsican/leaf#325] [Ideas]
It would be good to provide an API for retrieving runtime logs for diagnostic purposes.

## Feature breakdown
- retrieve most recent unformatted plaintext logs
- real-time updates

## Potential implementation
For the API endpoint, WebSocket may be a good choice to transfer a large amount of data and real-time updates.
For the log collection, I can see two possible ways:

1. Prepare two fixed-sized buffers. Divert the logs to one of them, and swap when full. Every time a new connection comes in, we send back the buffers as a whole. Broadcast new logs to connected clients as well.
2. `tail -f` on the log file and send back the outputs. For console-based logs, assign temporary files as backing storage.

Any ideas or suggestions?

**Comment by eycorsican** [comment]
Why go through all these inconvenience when you can just tail on the log file? Reading a file is much easier than connecting to an API endpoint.

**Reply by bdbai** [reply]
1. The UI may not have access to the log file as the core do. WebUI is an example.
2. If the user chooses not to write logs to a file i.e. `logoutput = console`, there is no way for an external program to get the logs.
